src/solvers/dgmulti/flux_differencing.jl

```diff
@@ -4,125 +4,171 @@
+                               sparsity_pattern::Nothing, skip_index=(i,j)->false)
+
+  rows, cols = axes(A)
+  for i in rows
     u_i = u[i]
     du_i = du[i]
-    for j in rows
-      if !skip_index(i,j)
-          du_i = du_i + ATr[j,i] * volume_flux(u_i, u[j])
+    for j in cols
+      # This routine computes only the upper-triangular part of the hadamard sum (A .* F).
+      # We avoid computing the lower-triangular part, and instead accumulate those contributions
+      # while computing the upper-triangular part (using the fact that A is skew-symmetric and F
+      # is symmetric).
+      if j > i && !skip_index(i, j)
+          AF_ij = A[i,j] * volume_flux(u_i, u[j], orientation, equations)
+          du_i = du_i + AF_ij
+          du[j] = du[j] - AF_ij
       end
     end
     du[i] = du_i
   end
 end
 
-# Version of function without `skip_index`, as mentioned in
-# https://github.com/trixi-framework/Trixi.jl/pull/695/files#r670403097
-@inline function hadamard_sum_A_transposed!(du, ATr, volume_flux, u)
-  rows, cols = axes(ATr)
-  for i in cols
+@inline function hadamard_sum!(du, A, volume_flux, orientation, u,
+                               equations, sparsity_pattern::AbstractSparseMatrix{Bool})
+  n = size(sparsity_pattern, 2)
+  rows = rowvals(sparsity_pattern)
+  for i in 1:n
     u_i = u[i]
     du_i = du[i]
-    for j in rows
-      du_i = du_i + ATr[j,i] * volume_flux(u_i, u[j])
+    for id in nzrange(sparsity_pattern, i)
+      j = rows[id]
+      if j > i
+        AF_ij = A[i,j] * volume_flux(u_i, u[j], orientation, equations)
+        du_i = du_i + AF_ij
+        du[j] = du[j] - AF_ij
+      end
     end
     du[i] = du_i
   end
 end
 
 # For DGMulti implementations, we construct "physical" differentiation operators by taking linear
 # combinations of reference differentiation operators scaled by geometric change of variables terms.
-# We use LazyArrays.jl for lazy evaluation of physical differentiation operators, so that we can
-# compute linear combinations of differentiation operators on-the-fly in an allocation-free manner.
+# We use a lazy evaluation of physical differentiation operators, so that we can compute linear
+# combinations of differentiation operators on-the-fly in an allocation-free manner.
 function build_lazy_physical_derivative(element, orientation,
-                                        mesh::VertexMappedMesh{2, Tri}, dg, cache)
-  @unpack Qrst_skew_Tr = cache
+                                        mesh::VertexMappedMesh{2}, dg, cache)
+  @unpack Qrst_skew = cache
   @unpack rxJ, sxJ, ryJ, syJ = mesh.md
-  QrskewTr, QsskewTr = Qrst_skew_Tr
   if orientation == 1
-    return LazyArray(@~ @. 2 * (rxJ[1,element] * QrskewTr + sxJ[1,element] * QsskewTr))
-  else
-    return LazyArray(@~ @. 2 * (ryJ[1,element] * QrskewTr + syJ[1,element] * QsskewTr))
+    return LazyMatrixLinearCombo(Qrst_skew, 2 .* (rxJ[1,element], sxJ[1,element]))
+  elseif orientation == 2
```

Comment:
```suggestion
  else # if orientation == 2
```
This avoids type instabilities when constant propagation fails.